### PR TITLE
DS-263 Add python standardization and IPA pipe

### PR DESCRIPTION
**Pull request by zeryx**
Required for new IPA algorithm architecture standardization
Also unlocks anaconda3 support

### CODE REVIEW

**Comment by dmorris-algo** [comment]
Since I don't have the context surrounding:
> new IPA algorithm architecture standardization

I'll defer to @anowell for final approval.

**Comment by zeryx** [comment]
> LGTM, only Q is in serve(self), why "for line in sys.stdin"? Shouldn't always only be 1 line?

nope! this pipe will remain open for the duration of the container, we only send EOF to the stdin pipe when the container is expected to exit, therefore the pause on new line acts as a wait loop.

File: Algorithmia/handler.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+        response_string = json.dumps({
+            'result': response,
+            'metadata': {
+                'content_type': content_type
+            }
+        })
+        return response_string
+
+    def write_to_pipe(self, data_string):
+        with open(self.FIFO_PATH, 'w') as f:
```
Comment (dmorris-algo) [inline Algorithmia/handler.py]: Can we open the FIFO once in the constructor vs. opening a new file descriptor every time we write to it?

File: Algorithmia/handler.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+            for line in sys.stdin:
+                request = json.loads(line)
+                formatted_input = self.format_data(request)
+                if load_result:
+                    apply_result = self.apply_func(formatted_input, load_result)
+                else:
+                    apply_result = self.apply_func(formatted_input)
+                formatted_response = self.format_response(apply_result)
+                self.write_to_pipe(formatted_response)
+        except Exception as e:
```
Comment (dmorris-algo) [inline Algorithmia/handler.py]: Per the [langpack guide's `bin/pipe` requirements](https://github.com/algorithmiaio/langpacks/blob/master/langpack_guide.md), I don't believe we want to terminate when an exception is thrown.

It appears as though the Algorithm Handler here will terminate on the first exception thrown.

File: Algorithmia/handler.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+import base64
+import traceback
+import six
+
+
+class Handler(object):
+
+    def __init__(self, apply_func, load_func=None):
+        """
+        Creates the handler object
```
Comment (dmorris-algo) [inline Algorithmia/handler.py]: This documentation is great. Is there a way to assert on the arity of the functions in Python, and return a descriptive error message, such that we could fail fast if I attempt to construct a `Handler` with a `load_func` that takes an argument.

File: Algorithmia/handler.py (hunk 1 above)
Comment (zeryx) [reply Algorithmia/handler.py]: there are some order of operations issues with this that only crop up during testing. I kept this as a strict refactor from the original pipe defined in langpacks, we could adjust this if we decide tho.

File: Algorithmia/handler.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+        :param apply_func: A required function that can have an arity of 1-2, depending on if loading occurs
+        :param load_func: An optional supplier function used if load time events are required, has an arity of 0.
+        """
+        self.FIFO_PATH = "/tmp/algoout"
+        self.apply_func = apply_func
+        if load_func:
+            self.load_func = load_func
+        else:
+            self.load_func = lambda: None
```
Comment (dmorris-algo) [inline Algorithmia/handler.py]: Minor nit: could we just do `def __init__(self, apply_func, load_func=lambda: None):`

File: Algorithmia/handler.py (hunk 1 above)
Comment (dmorris-algo) [reply Algorithmia/handler.py]: That's a reasonable decision given the heritage. I'm not too hung up on it so I think it's fine to leave it as-is with that in mind.

**Pull request merged**